### ISSUE TITLE
Phase 1 – Refactor GEMMA explainability into Producer–Consumer pipeline & fix SHAP crash

### ISSUE THREAD

**Issue by SefaGH** [SefaGH/bearish-alpha-bot#396]
## Context

Our `train-models.yml` workflow currently trains GEMMA models and, if they fail the balanced accuracy threshold, runs a post-training explainability step:

* Workflow: `.github/workflows/train-models.yml`
* Step 8 calls:

```yaml
python scripts/analyze_model_explainability.py \
  --model-path data/models/final/gemma_price.pt \
  --data-path data/cache/BTC-USDT_training_data.npz \
  --metadata-path features/gemma/metadata/feature_metadata.json \
  --output-dir ./analysis_artifacts
```

The explainability script is:
`scripts/analyze_model_explainability.py`

### Current problems

1. **SHAP crash (critical):**

   In `run_shap_analysis` (inside `analyze_model_explainability.py`), we see a runtime error in CI logs:

   ```text
   ❌ SHAP analizi sırasında hata oluştu: 'numpy.ndarray' object has no attribute 'cpu'
   ```

   The code assumes SHAP returns a PyTorch tensor:

   ```python
   shap_values_np = shap_values_for_error_class.cpu().numpy()
   ```

   But `shap_values_for_error_class` is already a NumPy array in practice, so `.cpu()` doesn’t exist.

2. **“Taklit eden” (duplicated) data prep in the analysis script:**

   `scripts/analyze_model_explainability.py` currently tries to **rebuild** the training pipeline:

   * Loads raw 87-feature data from `data/cache/BTC-USDT_training_data.npz`
   * Loads feature selection mask from `data/cache/gemma/feature_selection_mask.npy`
   * Applies the mask itself
   * Runs its own `train_test_split`
   * Loads and applies scaler again

   This is fragile: any deviation between training code (`src/ml/model_trainer.py` / `scripts/train_all_models.py`) and analysis code makes the confusion matrix and SHAP plots **no longer reflect the real test set** used to compute training metrics.

3. **Feature metadata mismatch / generic feature names:**

   * Analyzer expects feature names from `features/gemma/metadata/feature_metadata.json`, under a key like `selected_features`.
   * In reality, the **selected feature list** lives in:

     ```text
     features/gemma/selected/gemma_price_selected_82.json
     features/gemma/selected/gemma_regime_selected_82.json
     ```

     where feature names are under `features`.

   Result: the analysis step falls back to generic names `feature_0, feature_1…` and the plots become hard to interpret.

---

## Goal

Refactor the GEMMA explainability pipeline into a robust **Producer–Consumer** pattern:

* **Producer (trainer):**
  After training each GEMMA model, write the actual scaled train/test arrays and labels used for final metrics to a `.npz` file.

* **Consumer (analyzer):**
  Stop rebuilding the dataset. Instead, load that `.npz` file + the selected feature names JSON and run:

  * permutation importance,
  * confusion matrix & error cluster detection,
  * SHAP analysis

  **directly** on the real test set.

And at the same time:

* Fix the SHAP `.cpu()` crash in a type-safe way.
* Ensure feature names in plots come from the selected-features JSON.

---

## Scope

Affects:

* `src/ml/model_trainer.py` (or equivalent trainer used for GEMMA models)
* `scripts/train_all_models.py` (if it orchestrates GEMMA training)
* `scripts/analyze_model_explainability.py`
* `.github/workflows/train-models.yml`
* Optionally: feature metadata files under `features/gemma/...` for consistent naming

---

### Task 1 – Export analysis dataset from trainer (Producer)

**Files to touch:**

* `src/ml/model_trainer.py` (or the module that implements GEMMA `train_and_evaluate`)
* Possibly `scripts/train_all_models.py` (where GEMMA models are invoked)

**What to do:**

1. Locate the function that trains a GEMMA model and computes final metrics. This is likely something like:

   ```python
   def train_and_evaluate(..., model_name: str, ...):
       ...
       # X_train_scaled, X_test_scaled, y_train, y_test are available here
       # metrics dict is computed here
       return metrics
   ```

2. After you compute `X_train_scaled`, `X_test_scaled`, `y_test` (and ideally `X_train_scaled` for SHAP background), **save them to a compressed npz**:

   ```python
   import numpy as np
   from pathlib import Path
   import logging

   logger = logging.getLogger(__name__)

   # model_name is expected to be something like "gemma_price" or "gemma_regime"
   analysis_data_path = Path("data/cache") / f"{model_name}_analysis_test_data.npz"

   try:
       np.savez_compressed(
           analysis_data_path,
           X_train_scaled=X_train_scaled,
           X_test_scaled=X_test_scaled,
           y_test=y_test,
       )
       logger.info(f"✅ Saved analysis dataset for {model_name} to {analysis_data_path}")
   except Exception as e:
       logger.warning(f"⚠️ Failed to save analysis dataset for {model_name}: {e}")
   ```

3. Ensure `model_name` (e.g. `"gemma_price"` / `"gemma_regime"`) is passed into this function or otherwise available so that we can create:

   ```text
   data/cache/gemma_price_analysis_test_data.npz
   data/cache/gemma_regime_analysis_test_data.npz
   ```

4. Keep the existing metrics logging unchanged. This task only adds the `.npz` export.

---

### Task 2 – Simplify `analyze_model_explainability.py` to consume exported data (Consumer)

**File:**

* `scripts/analyze_model_explainability.py`

Currently this script:

* Takes `--data-path` (raw training data),
* Reads `feature_selection_mask.npy`,
* Performs its own train/test split and scaling,
* Tries to read feature names from `feature_metadata.json`.

We want to **stop doing that** and instead:

* Take `--analysis-data-path` (the new `.npz`),
* Take `--feature-names-path` (selected features JSON),
* Use those directly.

#### 2.1. Update CLI arguments

In `main()`:

* Replace:

  ```python
  parser.add_argument('--data-path', required=True, help="Path to training data .npz")
  parser.add_argument('--metadata-path', required=True, help="Path to feature metadata JSON")
  ```

* With something like:

  ```python
  parser.add_argument(
      '--analysis-data-path',
      required=True,
      help="Path to .npz file exported by the trainer (X_train_scaled, X_test_scaled, y_test). "
           "Example: data/cache/gemma_price_analysis_test_data.npz"
  )
  parser.add_argument(
      '--feature-names-path',
      required=True,
      help="Path to JSON containing selected feature names. "
           "Example: features/gemma/selected/gemma_price_selected_82.json"
  )
  ```

Keep `--model-path` and `--output-dir` as is.

> Note: If we still want `--metadata-path` for other purposes, we can keep it, but the main source of feature names should become `--feature-names-path`.

#### 2.2. Implement a simple `load_feature_names()` helper

Add near the top of the file:

```python
import json
from typing import List

def load_feature_names(feature_names_path: str) -> List[str]:
    """Load selected feature names from a JSON file.

    Expected format:
      {
        "features": ["feat_a", "feat_b", ...]
      }
    (as in features/gemma/selected/gemma_price_selected_82.json)
    """
    print(f"Özellik isimleri JSON dosyasından yükleniyor: {feature_names_path}")
    with open(feature_names_path, "r", encoding="utf-8") as f:
        data = json.load(f)

    if "features" not in data or not isinstance(data["features"], list):
        raise KeyError(f"'features' key not found or invalid in {feature_names_path}")

    names = data["features"]
    print(f"   ... {len(names)} adet özellik ismi yüklendi.")
    return names
```

We no longer need the old `load_data_and_features(data_path, metadata_path)` to load raw X and do masking. It can be removed or reduced to a wrapper that just calls `load_feature_names()` if still referenced.

#### 2.3. Load `.npz` instead of rebuilding train/test

In `main()`, **replace the whole block** that:

* loads `args.data_path` via `np.load`,
* loads `feature_selection_mask.npy`,
* applies mask,
* does `train_test_split`,
* calls `load_and_scale_data(X_train, X_test)`,

with something like:

```python
import numpy as np
from pathlib import Path

# ...

print(f"Nihai analiz verisi yükleniyor: {args.analysis_data_path}")
try:
    npz_data = np.load(args.analysis_data_path)
    X_train_scaled = npz_data["X_train_scaled"]
    X_test_scaled = npz_data["X_test_scaled"]
    y_test = npz_data["y_test"]
    print(f"   ... X_train_scaled shape: {X_train_scaled.shape}")
    print(f"   ... X_test_scaled shape: {X_test_scaled.shape}")
    print(f"   ... y_test length: {y_test.shape[0]}")
except Exception as e:
    print(f"HATA: Analiz verisi (.npz) yüklenemedi: {e}")
    sys.exit(1)

feature_names = load_feature_names(args.feature_names_path)

if X_test_scaled.shape[1] != len(feature_names):
    print(
        f"HATA: Veri sütun sayısı ({X_test_scaled.shape[1]}) ile "
        f"özellik isimleri ({len(feature_names)}) uyuşmuyor!"
    )
    sys.exit(1)
```

After this, you can keep the existing calls to:

* `run_permutation_importance(model_wrapper, X_test_scaled, y_test, feature_names, output_dir)`
* `run_shap_analysis(model_wrapper, X_train_scaled, X_test_scaled, y_test, y_pred, feature_names, output_dir)`

Just make sure they now use `X_train_scaled` / `X_test_scaled` loaded from `.npz`.

#### 2.4. Remove obsolete data prep code

* Delete or comment out the old code paths that:

  * Load `args.data_path` as raw 87-feature data,
  * Load and apply `feature_selection_mask.npy`,
  * Perform `train_test_split`,
  * Call `load_and_scale_data()`.

From now on, **all** analysis should be based on the trainer-exported `.npz`.

---

### Task 3 – Fix SHAP `.cpu()` crash in `run_shap_analysis`

**File:**

* `scripts/analyze_model_explainability.py`

Inside `run_shap_analysis(...)`, there is a block that converts SHAP values to NumPy:

```python
# shap.summary_plot bir PyTorch Tensörünü kabul etmez, NumPy dizisi bekler.
shap_values_np = shap_values_for_error_class.cpu().numpy()
```

This is what causes:

```text
'numpy.ndarray' object has no attribute 'cpu'
```

**Replace it** with a type-safe version:

```python
import numpy as np  # ensure this is imported at the top

# ...

# shap.summary_plot NumPy array bekler; hem tensor hem ndarray'i güvenli şekilde destekle
if hasattr(shap_values_for_error_class, "detach") and hasattr(shap_values_for_error_class, "cpu"):
    shap_values_np = shap_values_for_error_class.detach().cpu().numpy()
else:
    shap_values_np = np.array(shap_values_for_error_class)
```

This will work whether SHAP returns a tensor or a NumPy array.

---

### Task 4 – Wire up workflow to new arguments

**File:**

* `.github/workflows/train-models.yml`

In step:

```yaml
- name: 8. Hata Durumunda Açıklanabilirlik Analizi Çalıştırma
  if: steps.deploy_check.outputs.all_models_passed == 'false'
  run: |
    python scripts/analyze_model_explainability.py \
      --model-path data/models/final/gemma_price.pt \
      --data-path data/cache/BTC-USDT_training_data.npz \
      --metadata-path features/gemma/metadata/feature_metadata.json \
      --output-dir ./analysis_artifacts
```

Replace the arguments to use the new producer–consumer setup. For GEMMA_PRICE, assuming trainer writes:

* `data/cache/gemma_price_analysis_test_data.npz`
* and we already have: `features/gemma/selected/gemma_price_selected_82.json`

Update as:

```yaml
python scripts/analyze_model_explainability.py \
  --model-path data/models/final/gemma_price.pt \
  --analysis-data-path data/cache/gemma_price_analysis_test_data.npz \
  --feature-names-path features/gemma/selected/gemma_price_selected_82.json \
  --output-dir ./analysis_artifacts
```

If we later want to analyze `gemma_regime` as well, we can extend this call or add a second one with `gemma_regime_analysis_test_data.npz` & `gemma_regime_selected_82.json`.

---

## Acceptance Criteria

1. **No SHAP crash in CI:**

   * When `gemma_price` (and/or `gemma_regime`) fails the balanced accuracy threshold, step 8 in `train-models.yml` runs successfully.
   * The SHAP analysis step no longer raises `'numpy.ndarray' object has no attribute 'cpu'`.

2. **Producer–Consumer path is used:**

   * Logs from training show lines like:

     ```text
     ✅ Saved analysis dataset for gemma_price to data/cache/gemma_price_analysis_test_data.npz
     ```

   * Logs from analysis show that it loads `X_train_scaled`, `X_test_scaled`, `y_test` from that `.npz`, not from `BTC-USDT_training_data.npz`.

   * No `train_test_split` or scaler fitting happens inside `analyze_model_explainability.py` anymore.

3. **Feature names are meaningful:**

   * Permutation importance and SHAP plots reference actual feature names from `features/gemma/selected/gemma_price_selected_82.json` (e.g. `rsi_14`, `ema_50`, etc.), not `feature_0`, `feature_1`, ...
   * If feature count in the `.npz` and in the JSON does not match, the script fails fast with a clear error.

4. **Artifacts are generated as before (but more correct):**

   * `analysis_artifacts/` contains:

     * permutation importance plot,
     * confusion matrix / error report for the worst error cluster,
     * SHAP plots for misclassified examples.
   * These now reflect the **exact test set** used for final training metrics.

---

**Comment by SefaGH** [comment]
#### **Critical Action: Environment Setup and Verification**

1.  **Enforce Python Version 3.11:**
    *   The project requires Python `3.11.x`. The runner environment may default to a different version (e.g., 3.12).
    *   **Action:** Your first step must be to check the active Python version (`python --version`). If it is not `3.11`, you must stop, install Python 3.11, and switch the environment to use it exclusively.

2.  **Mandatory Full Dependency Installation:**
    *   The validation requires all project dependencies, including heavy libraries like `torch`, `pandas`, and `scikit-learn`. Skipping any package will lead to invalid test results.
    *   **Action:** After ensuring Python 3.11 is active, perform a complete and non-simplified installation of all packages listed in the project's dependency files (e.g., `requirements.txt`). Do not use mocked libraries for this validation.

---

#### **Execution Directive**

Once the environment is confirmed to be running **Python 3.11** and **all dependencies are fully installed**, proceed with executing the complete validation plan as detailed in the issue description above, starting from `Task 1`.

The goal is a full, non-simplified validation of the artifacts and code.
